### PR TITLE
feat: add tinlake-bot

### PR DESCRIPTION
**Pull request by offerijns**

### CODE REVIEW

File: tinlake-bot/src/tasks/closePools.ts [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+import { formatEvents, parseRatio } from '../util/formatEvents'
+import { PoolMap } from '../util/ipfs'
+import { pushNotificationToSlack } from '../util/slack'
+
+const e18 = new BN('10').pow(new BN('18'))
+
+export const closePools = async (pools: PoolMap, provider: ethers.providers.Provider, signer: ethers.Signer) => {
+  console.log('Checking if any pools can be closed')
+  for (let pool of Object.values(pools)) {
+    if (!pool.addresses) continue
```
Comment (philipstanislaus) [inline tinlake-bot/src/tasks/closePools.ts:14]: Can we wrap this for loop in a try block such that we can continue processing of other pools, even if one of them throws an error?

File: tinlake-bot/src/tasks/closePools.ts [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+      await tinlake.getTransactionReceipt(solveTx)
+
+      const currentTinRatio = parseRatio(await tinlake.getCurrentJuniorRatio())
+      pushNotificationToSlack(
+        `I just closed epoch ${id} for *<${config.tinlakeUiHost}pool/${pool.addresses.ROOT_CONTRACT}/${pool.metadata.slug}|${name}>*.`,
+        formatEvents(epochState, orders, false, currentTinRatio),
+        {
+          title: 'View on Etherscan',
+          url: `https://kovan.etherscan.io/tx/${solveTx.hash}`,
+        }
```
Comment (philipstanislaus) [inline tinlake-bot/src/tasks/closePools.ts:54]: Put this url in env?

File: tinlake-bot/src/tasks/submitSolutions.ts [hunk 3]
```diff
@@ -0,0 +1,31 @@
+import Tinlake from '@centrifuge/tinlake-js'
+import { ethers } from 'ethers'
+import { PoolMap } from '../util/ipfs'
+
+export const submitSolutions = async (pools: PoolMap, provider: ethers.providers.Provider, signer: ethers.Signer) => {
+  console.log('Checking if any solutions can be submitted')
+  for (let pool of Object.values(pools)) {
+    if (!pool.addresses) continue
```
Comment (philipstanislaus) [inline tinlake-bot/src/tasks/submitSolutions.ts:8]: Again, error handling?

File: tinlake-bot/src/tasks/executePools.ts [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+import config from '../config'
+import { parseRatio } from '../util/formatEvents'
+import { PoolMap } from '../util/ipfs'
+import { pushNotificationToSlack } from '../util/slack'
+
+export const executePools = async (pools: PoolMap, provider: ethers.providers.Provider, signer: ethers.Signer) => {
+  console.log('Checking if any pools can be executed')
+
+  for (let pool of Object.values(pools)) {
+    if (!pool.addresses) continue
```
Comment (philipstanislaus) [inline tinlake-bot/src/tasks/executePools.ts:13]: Again, try/catch loop?

File: tinlake.js/src/actions/coordinator.ts [hunk 5]
```diff
@@ -1,25 +1,44 @@
 import BN from 'bn.js'
-import { calculateOptimalSolution, Orders, SolverWeights, State } from '../services/solver/solver'
+import { calculateOptimalSolution, Orders, SolverResult, SolverWeights, State } from '../services/solver/solver'
 import { Constructor, PendingTransaction, TinlakeParams } from '../Tinlake'
 const web3 = require('web3-utils')
 
 export function CoordinatorActions<ActionsBase extends Constructor<TinlakeParams>>(Base: ActionsBase) {
   return class extends Base implements ICoordinatorActions {
-    getEpochState = async () => {
+    getEpochState = async (beforeClosing?: boolean) => {
```
Comment (philipstanislaus) [inline tinlake.js/src/actions/coordinator.ts:8]: Could you add a jsdoc comment here and below explaining what that variable does? Thanks!

**Review by philipstanislaus — APPROVED** [review]
Nice 🎉

File: tinlake-bot/src/index.ts [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+
+const provider = new ethers.providers.JsonRpcProvider(config.rpcUrl)
+let pools: PoolMap = {}
+
+const run = async () => {
+  console.log('Decrypting wallet')
+  const signer = await ethers.Wallet.fromEncryptedJson(config.signerEncryptedJson, config.signerPassword)
+  const signerWithProvider = signer.connect(provider)
+
+  console.log(`Booting Dennis 2.0 as ${signer.address}`)
```
[the remark this answers is not in the document]
Comment (offerijns) [reply tinlake-bot/src/index.ts:19]: @philipstanislaus everything run on k8s pods is logged automatically on GKE, so you can just look up the container logs there.

File: tinlake-bot/src/tasks/closePools.ts (hunk 2 above)
Comment (offerijns) [reply tinlake-bot/src/tasks/closePools.ts:54]: Ah that's actually a bug and something I was meaning to do, thanks!

File: tinlake.js/src/actions/coordinator.ts (hunk 5 above)
Comment (offerijns) [reply tinlake.js/src/actions/coordinator.ts:8]: It's actually explained a few lines below this, but you're right that it makes more sense to add it on top of the function, I'll move it.

File: tinlake-bot/src/index.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (philipstanislaus) [reply tinlake-bot/src/index.ts:18]: Awesome 👍

**Pull request merged**